### ISSUE TITLE
Dynamic water (plane) level

### ISSUE THREAD

**Issue by sprunk** [beyond-all-reason/spring#1446]
Bill of materials to be done before starting this ticket:
* #1445

----

It would be good if the water plane height could be adjusted dynamically. This is a pretty common request but has historically been considered difficult, both to implement in engine and to workaround game-side. Currently to raise the water level by X you have to adjust the entire terrain by -X instead, which comes with its own caveats.

A water plane level getter function exists but is not used comprehensively (see #1445). The core of this ticket is a request is to pair it up with a setter, initially just for the plane, and of course expose it to Lua.

**Comment by sprunk** [comment]
Down the line the "always flat plane" limitation would be good to remove as well. It is way out of the scope of this ticket but it is good to keep it in mind in case it crops up when designing the implementation. The two approaches would be
* water could have a simulated flow. malric from ZK made a tech demo once ([ZK forum thread](https://zero-k.info/Forum/Thread/29599), [youtube vid](https://www.youtube.com/watch?v=2CMk95pbqsM), even [better YT vid](https://www.youtube.com/watch?v=cQW8WXNpYXk)) but I don't know if there is source anywhere. Also I think it's an impractical gimmick.
* water could be defined by a separate heightmap (allows having rivers flowing from the mountains etc). This seems both simpler and more practical, though deformation would be a question (what if you flatten the mountain?).

**Comment by jacobguenther** [comment]
> water could have a simulated flow

If somebody wants to implement that then I would recommend _Shallow Water Equations in Real-Time Computer Graphics_, Peeter Parna, a 300 page thesis on water simulation with references to nearly every other paper on the subject.


> It would be good if the water plane height could be adjusted dynamically

When I get back I could look into this a bit.

**Comment by jacobguenther** [comment]
One problem I see with dynamic water height is that some maps new and old will paint the the water floor blue. Which should really be handled by the terrain shader(with settings that maps can use) not map textures.

**Comment by TheGrueColor** [comment]
>     * water could have a simulated flow. malric from ZK made a tech demo once ([ZK forum thread](https://zero-k.info/Forum/Thread/29599), [youtube vid](https://www.youtube.com/watch?v=2CMk95pbqsM), even [better YT vid](https://www.youtube.com/watch?v=cQW8WXNpYXk)) but I don't know if there is source anywhere. Also I think it's an impractical gimmick.
Source code of malric's dynamic water:
https://github.com/vladmihaisima/spring/tree/water1

**Comment by jacobguenther** [comment]
I have questions like what should happen to a water unit(boat/sub) that should get beached or a land unit that ends up in the water? How would a changing water plane affect pathfinding? How expensive would it be.

**Comment by lostsquirrel1** [comment]
Units that find them,selves outside their terrain movement tolerance levels would be unable to move, but would then move fine if the situation changes to return them back within their tolerances.

It can be expensive for the pathing system. We can write some new optimizations to limit the impact, but the real issue is that the pathing system updates over time and if too much changes in one go then it will take time before the pathing system has fully caught up - it can take 20-30 seconds in-game time for example for wide map changes. We can mitigate by keeping water level changes small and gradual over time.

**Comment by lostsquirrel1** [comment]
Unlike other RTS games our pathing system does not rebuild the whole path map every frame. In BAR for example, we have 43 path maps with dense terrain data [a tank typically covers 25 squares], where RTS games typically have 1 path map and a typical tank covers 1 square.

**Comment by jacobguenther** [comment]
Is there a specific paper that describes the general way that BAR does it? I'd like to look into this a little.

**Comment by lostsquirrel1** [comment]
No. I've been meaning to write up an article series covering this, but I haven't found the time. It would certainly be helpful because there's a lot going on.

**Comment by jacobguenther** [comment]
I look forward to the day.

**Comment by drivver44** [comment]
> I have questions like what should happen to a water unit(boat/sub) that should get beached or a land unit that ends up in the water? How would a changing water plane affect pathfinding? How expensive would it be.

for `qtpfs` it wouldnt be any different to recalculating an entire movedef pathing on a map (fairly expensive and slowish and batched when possible)
subs have a height that they go to in bar and zk.

**Comment by sprunk** [comment]
> If somebody wants to implement that then I would recommend Shallow Water Equations in Real-Time Computer Graphics
> Source code of malric's dynamic water

Thanks. In the scope of this ticket, I am only mentioning dynamic water (and heightmap style water) in the way of "please keep in mind that this is one of the possible directions after adjustable waterplane when designing stuff".

> what should happen to a water unit(boat/sub) that should get beached or a land unit that ends up in the water? 

Probably the same thing that happens if you "adjust waterplane" the currently available workaround method (i.e. by adjusting the entire terrain in the opposite direction).

> It can be expensive for the pathing system. We can write some new optimizations to limit the impact
> We can mitigate by keeping water level changes small and gradual over time.

Yes, I think an initial implementation doesn't need to worry about performance as whether something is too slow or not will depend on how a game uses it. For the scope of this ticket I think just adding the basic capability to adjust the plane at all would be sufficient.
